File: 3semestre/poo/runcodes/Tabuleiro/src/Tabuleiro.java
public class Tabuleiro {
    private int[][] tabela;
    private int l_vazia, c_vazia; //posição da casa vazia
    private int tamanho; //tamanho do tabuleiro

    public Tabuleiro(int[] numeros) {

        tamanho = (int) Math.sqrt(numeros.length) ; 
        tabela = new int[tamanho][tamanho];
        int count = 0;
        for (int i = 0; i < tamanho; i++) {
            for (int j = 0; j < tamanho; j++) {
                tabela[i][j] = numeros[count];
                count++;
                if (this.tabela[i][j] == 0) {
                    l_vazia = i;
                    c_vazia = j;
                }
            }
        }

    }

    public void printa_tabuleiro() {
        for (int t = 0; t < tamanho; t++) { //"tampa" da tabela
            System.out.print("+------");
        }
        for (int i = 0; i < tamanho; i++) {

            System.out.print("+\n|");

            for (int j = 0; j < tamanho; j++) {
                if(tabela[i][j] < 10)
                    System.out.print(" "); //correção caso numero<10

                System.out.printf("  %d  |", tabela[i][j]);
            }
            System.out.println();
            for (int t = 0; t < tamanho; t++) { //printa em baixo dos numeros
                System.out.print("+------");
            }
        }
        System.out.print("+\n\n");
    }
    public void move_tabuleiro(String op) {
        int l_update = 0, c_update = 0;
        for (int count = 0; count < op.length(); count++) {
            switch (op.charAt(count)) {
                case 'u': {
                    l_update = 1;
                    c_update = 0;
                    break;
                }
                case 'd': {
                    l_update = -1;
                    c_update = 0;
                    break;
                }
                case 'l': {
                    l_update = 0;
                    c_update = 1;
                    break;
                }
                case 'r': {
                    l_update = 0;
                    c_update = -1;
                    break;
                }
                default: System.out.println("Operação não reconhecida");
            }

            if (l_vazia+l_update >= 0 && l_vazia+l_update < tamanho //verifica validade do movimento
            && c_vazia+c_update < tamanho && c_vazia+c_update >= 0) {
                this.tabela[l_vazia][c_vazia] = this.tabela[l_vazia + l_update][c_vazia + c_update]; //swap
                this.tabela[l_vazia + l_update][c_vazia + c_update] = 0;

                c_vazia += c_update; //att a posição vazia
                l_vazia += l_update;
            }
            printa_tabuleiro();
        }
    }

    public boolean valida_tabuleiro() {
        int count = 0;
        for (int i = 0; i < tamanho; i++) {
            for (int j = 0; j < tamanho; j++) {
                if (this.tabela[i][j] != count) {
                    return false;
                }
                count++;
            }
        }
        return true;
    }
}

File: 3semestre/poo/tabuleiro_python/tabuleiro.py
import math
import sys
#import random form Random
#rd.seed()

class Tabuleiro:
    def __init__(self, numeros):
        self.tamanho = int(math.sqrt(len(numeros)))
        self.tabela = [[0 for _ in range(self.tamanho)] for _ in range(self.tamanho)]
        self.l_vazia = -1
        self.c_vazia = -1
    
        for i in range(self.tamanho):
            for j in range(self.tamanho):
                self.tabela[i][j] = numeros[i*self.tamanho+j]
                if (numeros[i*self.tamanho+j] == '0'):
                    self.l_vazia = i 
                    self.c_vazia = j
    
    def printa_tabuleiro(self):
        for _ in range(self.tamanho):
            sys.stdout.write("+------")

        for i in range(self.tamanho):
            sys.stdout.write("+\n|")

            for j in range(self.tamanho):
                if int(self.tabela[i][j]) < 10:
                    sys.stdout.write(" ") #correção caso numero<10
                
                if (self.tabela[i][j] == '0'):
                    sys.stdout.write("     |")
                else:
                    sys.stdout.write("  {}  |".format(self.tabela[i][j]))
            
            sys.stdout.write("\n")
            for t in range(self.tamanho):  #printa em baixo dos numeros
                sys.stdout.write("+------")
            
        
        sys.stdout.write("+\n\n")
    
    def move_tabuleiro(self, str_op):
        l_update = 0
        c_update = 0
        for char in str_op:
            match char:
                case 'u':
                    l_update = 1
                    c_update = 0

                case 'd': 
                    l_update = -1
                    c_update = 0
                
                case 'l':
                    l_update = 0;
                    c_update = 1;
                
                case 'r': 
                    l_update = 0
                    c_update = -1
                    
                case _:
                    print("Operacao invalida")
                
            l_nova = self.l_vazia + l_update
            c_nova = self.c_vazia + c_update            
    
            if (0 <= l_nova < self.tamanho and 0 <= c_nova < self.tamanho):
                self.tabela[self.l_vazia][self.c_vazia] = self.tabela[l_nova][c_nova]
                self.tabela[l_nova][c_nova] = '0'

                self.c_vazia += c_update # att a posição vazia
                self.l_vazia += l_update

            self.printa_tabuleiro()

    def valida_tabuleiro(self):
        for i in range(self.tamanho):
            for j in range(self.tamanho):
                if (self.tabela[i][j] != str(i*self.tamanho+j)):
                    return False
                
        return True